### PR TITLE
Allow using path suffixes to associate language file-types

### PR DESCRIPTION
**Pull request by midnightexigent**
Related: 

- #2443 
- #1426

After playing around with `shellcheck`, I think it's not the way to go mostly because of the path separator between linux and windows (\ vs /). Hence this PR uses globbing instead 

Now it's possible to associate for example `.git/config` to the `git-config` grammar by adding the following

```toml
[[language]]
name = "git-config"
file-globs = ["*.git/config"] 
```

image

### CODE REVIEW

**Comment by midnightexigent** [comment]
Also, I didn't manage to get the `ssh-client-config` to work. Anyone has any ideas.. ?

**Comment by sudormrfbin** [comment]
Do we really need file globbing here? Trying to match the end of the file path to a partial path given in `file-types` covers most of the cases that I can think of. So something like this:

```toml
[[language]]
name = "ssh-client-config"
scope = "source.ssh-client-config"
file-types = [ ".ssh/config" ]
```

We will definitely lose flexibility without globbing, so dropping it might not be a good idea anyway. A suggestion is to use the current `file-types` key instead of introducing a new `file-globs` key; i.e. `file-types` could itself understand glob patterns.

**Comment by midnightexigent** [comment]
@sudormrfbin 

I thought about those approaches before opting for globbing, here is what I ran into when considering those approaches

> Trying to match the end of the file path to a partial path given in `file-types` covers most of the cases that I can think of. So something like this:
> 
> ```toml
> [[language]]
> name = "ssh-client-config"
> scope = "source.ssh-client-config"
> file-types = [ ".ssh/config" ]
> ```

This runs into 2 problems: 

- Handling of the '/' in a cross-platform way (on windows, the path of the file will be .ssh\\config which won't match)
- `.ssh/config` is relative to what ? even if the file is opened from the home directory `hx .ssh/config`, the path will be its absolute path aka `/home/user/.ssh/config` which won't match with `.ssh/config` without some sort of wacky string comparison. The alternative is to do something like described in [comment](https://github.com/helix-editor/helix/issues/2443#issuecomment-1121109890) which matches the exact path enabling something like `file-types = ["~/.ssh/config"]` or `file-types = ["$root/.git/config"]`. But I also discarded that approach because I couldn't think of a clean way to do it in a cross-platform manner (`/` vs `\\`)

> A suggestion is to use the current `file-types` key instead of introducing a new `file-globs` key; i.e. `file-types` could itself understand glob patterns.

Handling each of the strings in the `file-types` array as a glob feels kind of awkward and might result in undesirable behavior in some cases. Handling globs separately felt more natural

**Comment by archseer** [comment]
I'm a bit wary of importing another crate for something this small too.

> * Handling of the '/' in a cross-platform way (on windows, the path of the file will be .ssh\config which won't match)

We can document we expect `/` in these strings, then replace `/` with https://doc.rust-lang.org/std/path/constant.MAIN_SEPARATOR.html internally

> * `.ssh/config` is relative to what

It doesn't have to be relative to anything. You'd take the absolute filepath and check for `ends_with?`

File: languages.toml [hunk 1]
```diff
@@ -946,8 +946,8 @@ source = { git = "https://github.com/tree-sitter/tree-sitter-regex", rev = "e1cf
 name = "git-config"
 scope = "source.gitconfig"
 roots = []
-# TODO: allow specifying file-types as a regex so we can read directory names (e.g. `.git/config`)
 file-types = [".gitmodules", ".gitconfig"]
+file-globs = ["*.git/config"]
```
Comment (pickfire) [inline languages.toml:950]: I think this is missing `~/.config/git/config` with this

**Comment by midnightexigent** [comment]
> I'm a bit wary of importing another crate for something this small too.
> 
> > * Handling of the '/' in a cross-platform way (on windows, the path of the file will be .ssh\config which won't match)
> 
> We can document we expect `/` in these strings, then replace `/` with https://doc.rust-lang.org/std/path/constant.MAIN_SEPARATOR.html internally
> 
> > * `.ssh/config` is relative to what
> 
> It doesn't have to be relative to anything. You'd take the absolute filepath and check for `ends_with?`

I guess that would work, the behavior sounds a bit like a subset of globbing tho. But ok, I understand your point of not wanting to add dependencies that could be easily implemented by hand. 

I'll update the PR soon

File: languages.toml [hunk 2]
```diff
@@ -1355,3 +1355,13 @@ indent = { tab-width = 2, unit = "  " }
 [[grammar]]
 name = "meson"
 source = { git = "https://github.com/bearcove/tree-sitter-meson", rev = "feea83be9225842490066522ced2d13eb9cce0bd" }
+
+[[language]]
+name = "sshclientconfig"
+scope = "source.sshclientconfig"
+file-types = [".ssh/config"]
```
Comment (pickfire) [inline languages.toml:1362]: There is also `/etc/ssh/ssh_config`

File: runtime/queries/sshclientconfig/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,324 @@
+(host) @keyword
+(host_value) @identifier
+
+(match) @keyword
+(match_value) @identifier
+
+(add_keys_to_agent) @keyword
+(add_keys_to_agent_value) @boolean
```
Comment (pickfire) [inline runtime/queries/sshclientconfig/highlights.scm:8]: I think this should be a separate pull request.

Also, it seemed kinda hard to manage this if we have to color each keyword manually, shouldn't it have be grouped?

Comment (midnightexigent) [reply runtime/queries/sshclientconfig/highlights.scm]: Opened #2498

**Review by dead10ck — COMMENTED** [review]
It does look like all of the comments were addressed. However, I can't help but think that the behavior of the `file-types` option is getting confusing and hard to reason about. If I'm understanding this right, the logic of the key is now:

1. File name is an exact match
2. File extension is an exact match
3. File path contains the value as a suffix

I guess it would be alright if this were documented, but I do think that having a separate key with its own semantics would be warranted too.

**Comment by pickfire** [comment]
Yeah, it seemed confusing to me too, some of the cases could even overlap. I think a better way is to make use of globbing, so that is easier for users to learn and don't have overlaps.

1. File name is an exact match - `file`
2. File extension is an exact match - `*.ext`
3. File path contains the value as a suffix - `**/*.ext` or `**.ext`
4. File path `.git/config`

**Comment by midnightexigent** [comment]
That is what I initially proposed in https://github.com/helix-editor/helix/pull/2455#issue-1231137605 but I ended up changing my mind because I agreed with https://github.com/helix-editor/helix/pull/2455#issuecomment-1123054205 because the other approaches seem to bring a lot more flexibility in order to handle use-cases that probably won't occur

>     1. File name is an exact match - `file`
> 
>     2. File extension is an exact match - `*.ext`
> 
>     3. File path contains the value as a suffix - `**/*.ext` or `**.ext`
> 
>     4. File path `.git/config`

1 and 2 are already what's implemented; 3 and 4 are essentially the same thing - there is no exact path matching since when you open let's say `hx .git/config`, its path is going to be `/home/user/Source/project/.git/config`, so you would need to match the suffix for it to be detected (which is what's implemented). 

3 approaches have been discussed so far: 

- with `shellexpand` 
```toml
[[language]]
name = "ssh-client-config"
file-types = ["~/.ssh/config"] 

[[language]]
name = "git-config"
file-types = ["~/.gitconfig", "${workspace}/.git/config"]
```
- globbing (note that we would use a attribute here)
```toml
[[language]]
name = "git-config"
file-globs = ["*.git/config"] 

[[language]]
name = "ssh-client-config"
file-globs = ["*.ssh/config"] 
```
- Manually doing suffix matching which essentially boils down to doing pseudo-globbing 

IMO, the objectively cleanest approach would be to separate each strategy and add a field like this 

```rs
enum FileMatchStrategy {
    ByName(String), 
    ByExtension(String),
    ByGlob(String),
    ByShellExpression(String),
}
struct LanguageConfig {
    //* fields *//
    file_matchers: Vec<FileMatchStrategy>,
}
```

Which would allow the user to configure what strategy takes precedence over the other in a flexible way (also this way, we could add strategies without breaking existing configs)

But again, the problem with this solution is that the use cases don't warrant that kind of flexibility

**Comment by dead10ck** [comment]
I do think being explicit about the strategy would make the config a lot easier to comprehend, especially to newcomers. With this addition, I can't help but feel like you'd have to look at the source code to completely understand how the option behaves.

**Review by the-mikedavis — COMMENTED** [review]
Sorry for the delay on this. This looks good as an incremental improvement to file detection. I'd like globbing in the long-run since Starlark files can also be detected by `BUILD.*` (like `BUILD.dep1`, `BUILD.dep2`. 🙃) but that can be tackled separately since it brings additional complexity and discussion.

As dead10ck says though, it would be a big improvement to document the precedence rules of `file-types` in `book/src/languages.md`. A blurb in after the table in the "## Language configuration" section would be ideal I think

**Comment by midnightexigent** [comment]
How about now ?

**Comment by dead10ck** [comment]
That looks good to me, thanks!

**Review by the-mikedavis — APPROVED** [review]
I think this looks great! 👍

File: book/src/languages.md [hunk 4]
```diff
@@ -50,7 +50,7 @@ These configuration keys are available:
 | `name`                | The name of the language                                      |
 | `scope`               | A string like `source.js` that identifies the language. Currently, we strive to match the scope names used by popular TextMate grammars and by the Linguist library. Usually `source.<name>` or `text.<name>` in case of markup languages |
 | `injection-regex`     | regex pattern that will be tested against a language name in order to determine whether this language should be used for a potential [language injection][treesitter-language-injection] site. |
-| `file-types`          | The filetypes of the language, for example `["yml", "yaml"]`. Extensions and full file names are supported.  |
+| `file-types`          | The filetypes of the language, for example `["yml", "yaml"]`. This attempts to match by exact file name, then by file extension, then by path suffix  |
```
Comment (pickfire) [inline book/src/languages.md:53]:
```suggestion
| `file-types`          | The filetypes of the language, for example `["yml", "yaml"]`. This attempts to match by exact file name, then by file extension, then by path suffix. |
```

File: book/src/languages.md [hunk 5]
```diff
@@ -62,6 +62,12 @@ These configuration keys are available:
 | `grammar`             | The tree-sitter grammar to use (defaults to the value of `name`) |
 | `formatter`           | The formatter for the language, it will take precedence over the lsp when defined. The formatter must be able to take the original file as input from stdin and write the formatted file to stdout |
 
+Additionnaly, entries in `file-types` are matched accoding to different strategies, in the following order :
+
+1. File name is an exact match -- example: `.zshrc`
+2. File extension is an exact match -- example: `toml`
+3. File path contains the value as a suffix -- example: `.git/config`
+
```
Comment (pickfire) [inline book/src/languages.md:70]: Unnecessary with the following suggestions above.

**Comment by pickfire** [comment]
I wish there is the glob thing here so it can be a prerequisite for https://github.com/helix-editor/helix/pull/2455

**Comment by pickfire** [comment]
I was doing a fallback approach on https://github.com/helix-editor/helix/pull/3249 since the mechanism I originally think of can be confusing, maybe you might want to take a look at that?

**Comment by midnightexigent** [comment]
So what's blocking here ?
